src/match.c

```diff
@@ -1850,9 +2012,57 @@ r_obj* expand_compact_indices(const int* v_o_haystack,
         v_out_needles[loc_out] = loc_needles + 1;
         v_out_haystack[loc_out] = loc_haystack + 1;
 
-        if (retain_remaining_haystack) {
-          // This haystack value was a match, so it isn't "remaining"
-          v_detect_remaining_haystack[loc_haystack] = 0;
+        if (track_matches_haystack) {
+          if (check_multiple_haystack) {
+            // `true` if a match already existed
+            any_multiple_haystack = v_detect_matches_haystack[loc_haystack];
+
+            if (any_multiple_haystack) {
+              loc_first_multiple_haystack = loc_haystack;
+
+              switch (relationship) {
+              case VCTRS_RELATIONSHIP_one_to_one:
+                stop_matches_relationship_one_to_one(
+                  loc_first_multiple_haystack,
+                  "haystack",
+                  needles_arg,
+                  haystack_arg,
+                  error_call
+                );
+              case VCTRS_RELATIONSHIP_one_to_many:
+                stop_matches_relationship_one_to_many(
+                  loc_first_multiple_haystack,
+                  needles_arg,
+                  haystack_arg,
+                  error_call
+                );
+              case VCTRS_RELATIONSHIP_warn_many_to_many: {
+                if (any_multiple_needles) {
+                  warn_matches_relationship_many_to_many(
+                    loc_first_multiple_needles,
+                    loc_first_multiple_haystack,
+                    needles_arg,
+                    haystack_arg,
+                    error_call
+                  );
+                }
+
+                // We know there are multiple and don't need to continue checking
+                check_multiple_haystack = false;
+
+                // Only continue tracking if needed for `remaining`
+                track_matches_haystack = retain_remaining_haystack;
+
+                break;
+              }
```

Comment:
Basic idea here is that as we process the `haystack` matches, we record the location in `haystack` that the match corresponds to in `v_detect_matches_haystack`. If we see that location more than once, that corresponds to a multiple match and we take the appropriate action based on `relationship`.

If the "action" is to just throw a warning, then we throw it once and after that we typically no longer need to check for multiple `haystack` matches.

The exception is if `remaining = "error" / <int>` is set, in which case we still have to track usage of all of the `haystack` locations to figure out which locations _weren't_ used at all (i.e. they are "unmatched" in dplyr join terminology). So that is what `track_matches_haystack = retain_remaining_haystack;` is for.